### DISCUSSION TITLE
About V300

### DISCUSSION THREAD

**Discussion by Northward712** [ergo-services/ergo#174] [General]
Hi, I saw the submission of the new V300 branch. Is there any relevant documentation to check?

**Comment by halturin** [comment]
Hi,
Its in progress now and will be published along with the release 3.0.

**Reply by Northward712** [reply]
When will version 3.0 be released?

**Comment by Northward712** [comment]
Hello, I’m currently trying to use the V300 branch for beta development. However, I noticed that the actor.Call function frequently times out during usage. I’m wondering if I’m using it incorrectly or if there might be another issue causing this problem. Could you please provide me with some guidance or assistance on this matter? Thank you.

________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年2月19日 2:42
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


Hi,
Its in progress now and will be published along with the release 3.0.

―

**Reply by halturin** [reply]
Could you please share the way you use it?

**Comment by Northward712** [comment]
func GetNodeById(process gen.Process, id string) (gen.Atom, bool, error) {
    reply, err := process.Call(gen.Atom("#cluster"), sync_search_node_by_id{id: id})
    if err != nil {
       return "", false, errors.WithMessage(err, "get node by id failed")
    }
    ret := reply.(sync_search_node_by_id_ret)
    return ret.node, ret.ok, nil
}

func CallById(process gen.Process, to any, id string, msg any) (any, error) {
    node, ok, err := GetNodeById(process, id)
    if err != nil {
       return nil, errors.WithMessage(err, "'CallById' get remote game node failed")
    }
    if !ok {
       return nil, ErrServerNotFound
    }
    fmt.Println(node)
    switch target := to.(type) {
    case gen.Alias:
       return process.CallAlias(target, msg, 5)
    case gen.PID:
       return process.CallPID(target, msg, 5)
    case string:
       return process.CallProcessID(gen.ProcessID{Name: gen.Atom(target), Node: node}, msg, 5)
    case gen.Atom:
       return process.CallProcessID(gen.ProcessID{Name: target, Node: node}, msg, 5)
    default:
       process.Log().Error("'SendById' unknown target type: %T", target)
       return nil, nil
    }
}

Just like this, when calling with the same node, there may be a timeout. When calling with a remote node, the other party does not fail to receive the call, but is unable to effectively return. However, returning “nil, errors.New(“xxx”)” will succeed in returning, but it will not terminate the blocking of the Call invocation.
________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 17:17
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


Could you please share the way you use it?

―

**Comment by halturin** [comment]
It is not allowed to use `gen.Process` outside of the actor and must be used within `Handle*` callbacks only (or the same goroutine)

**Comment by halturin** [comment]
here is an example you might find useful https://github.com/ergo-services/examples/tree/v300

**Comment by Northward712** [comment]
However, timeout can also occur when using it internally.
________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 17:58
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


here is an example you might find useful https://github.com/ergo-services/examples/tree/v300

―

**Comment by halturin** [comment]
>However, timeout can also occur when using it internally.

default timeout for Call requests is 5 seconds. Make sure your process (that handles this request) is able to handle it in time.

**Comment by Northward712** [comment]
like this


func (this *cluster) HandleMessage(from gen.PID, message any) error {
    switch val := message.(type) {
    case async_node_discover:
       node_name := gen.Atom(fmt.Sprintf("%s@%s", val.id, val.host))
       if val.id == dust.Id() {
          return nil
       }
       if err := dust.ConnectNode(node_name); err != nil {
          this.Log().Error("cluster discover new node %s but connect failed: %s", node_name, err)
          return nil
       }
       dict.HashPut(this.dict, dist_key, val.id, node_name, dict.Infinity)
       fmt.Println(this.CallProcessID(gen.ProcessID{Name: "#hub", Node: node_name}, "test call", 5))
    default:
       this.Log().Error("cluster unhandled async message from [%s]%s. message: %T --> %#v", from.Node, from, val, val)
    }
    return nil
}

________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 17:58
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


here is an example you might find useful https://github.com/ergo-services/examples/tree/v300

―

**Comment by Northward712** [comment]
According to the logs of another node, it appears that the recipient receives the call message immediately and responds, but the caller does not receive the response effectively.
________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 18:01
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


However, timeout can also occur when using it internally.

default timeout for Call requests is 5 seconds. Make sure your process (that handles this request) is able to handle it in time.

―

**Reply by halturin** [reply]
please, share your HandleCall on the remote process.

**Comment by Northward712** [comment]
[cid:1a2b9d7b-ca1a-4a06-a715-64501027708f]

this is my remote node HandleCall logic
________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 18:03
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


please, share your HandleCall on the remote process.

―

**Comment by Northward712** [comment]
func factor() gen.ProcessBehavior { return &adaptor{} }

type adaptor struct {
    act.Actor
}

func (this *adaptor) HandleMessage(from gen.PID, message any) error {
    switch val := message.(type) {

    default:
       this.Log().Error("adaptor got unhandled object from [%s]%s. object: %T --> %#v", from.Node, from, val, val)
       return nil
    }
}

func (this *adaptor) HandleCall(from gen.PID, ref gen.Ref, request any) (any, error) {
    this.Log().Warning("Actor.HandleCall: unhandled request from %s", from)
    return "test reply", nil
}

this is my remote node HandleCall logic
________________________________
发件人: Cui WenBin ***@***.***>
发送时间: 2024年3月5日 18:08
收件人: ergo-services/ergo ***@***.***>
主题: 回复: [ergo-services/ergo] About V300 (Discussion #174)

[cid:1a2b9d7b-ca1a-4a06-a715-64501027708f]

this is my remote node HandleCall logic
________________________________
发件人: Taras Halturin ***@***.***>
发送时间: 2024年3月5日 18:03
收件人: ergo-services/ergo ***@***.***>
抄送: Kirin ***@***.***>; Author ***@***.***>
主题: Re: [ergo-services/ergo] About V300 (Discussion #174)


please, share your HandleCall on the remote process.

―

**Comment by Northward712** [comment]
Thank you for your reply and answer. I need to check to see if there is something wrong with the way I am using it.

**Comment by halturin** [comment]
demo.go:
```golang
package main

import (
	"time"

	"ergo.services/ergo"
	"ergo.services/ergo/act"
	"ergo.services/ergo/gen"
	"ergo.services/ergo/net/edf"
	"ergo.services/logger/colored"
)

type doCallLocal struct{}
type doCallRemote struct{}

type MyMessage struct {
	MyBool   bool
	MyString string
}

func main() {
	var options gen.NodeOptions

	// disable default logger to get rid of multiple logging to the os.Stdout
	options.Log.DefaultLogger.Disable = true

	// add logger "colored".
	loggercolored, err := colored.CreateLogger(colored.Options{})
	if err != nil {
		panic(err)
	}
	options.Log.Loggers = append(options.Log.Loggers, gen.Logger{Name: "cl", Logger: loggercolored})

	options.Log.Level = gen.LogLevelInfo
	// options.Log.Level = gen.LogLevelTrace

	// set network cookie
	options.Network.Cookie = "123"
	// to be able to use self-signed cert
	options.Network.InsecureSkipVerify = true

	// starting node1
	node1, err := ergo.StartNode("node1@localhost", options)
	if err != nil {
		panic(err)
	}

	// register network messages
	if err := edf.RegisterTypeOf(MyMessage{}); err != nil {
		panic(err)
	}

	// use the same options, but remove loggers we added for node1 to use the default one in node2
	options.Log.Loggers = nil
	options.Log.DefaultLogger.Disable = false
	node2, err := ergo.StartNode("node2@localhost", options)
	if err != nil {
		panic(err)
	}
	defer node2.StopForce()

	node2.SpawnRegister("b", factoryB, gen.ProcessOptions{})

	node1.SpawnRegister("b", factoryB, gen.ProcessOptions{})
	node1.SpawnRegister("a", factoryA, gen.ProcessOptions{})

	node1.Send(gen.Atom("a"), doCallLocal{})

	node1.Wait()
}

//
// Actor A
//

func factoryA() gen.ProcessBehavior {
	return &actorA{}
}

type actorA struct {
	act.Actor
}

func (a *actorA) Init(args ...any) error {
	a.Log().Info("started A process on %s: %s", a.Name(), a.Node().Name())
	return nil
}

func (a *actorA) HandleMessage(from gen.PID, message any) error {
	switch message.(type) {
	case doCallLocal:
		local := gen.Atom("b")
		if result, err := a.Call(local, MyMessage{MyString: "abc"}); err == nil {
			a.Log().Info("received result: %#v", result)
		} else {
			a.Log().Error("call local process failed: %s", err)
		}
		a.SendAfter(a.PID(), doCallRemote{}, time.Second)
		return nil

	case doCallRemote:
		remote := gen.ProcessID{Name: "b", Node: "node2@localhost"}
		if result, err := a.Call(remote, MyMessage{MyBool: true, MyString: "def"}); err == nil {
			a.Log().Info("received result: %#v", result)
		} else {
			a.Log().Error("call remote process failed: %s", err)
		}

		a.SendAfter(a.PID(), doCallLocal{}, time.Second)
		return nil

	}

	a.Log().Error("unknown message %#v", message)
	return nil
}

//
// Actor B
//

func factoryB() gen.ProcessBehavior {
	return &actorB{}
}

type actorB struct {
	act.Actor
}

func (b *actorB) Init(args ...any) error {
	b.Log().Info("started B process on %s: %s", b.Name(), b.Node().Name())
	return nil
}

func (b *actorB) HandleCall(from gen.PID, ref gen.Ref, request any) (any, error) {
	switch r := request.(type) {
	case MyMessage:
		b.Log().Info("received MyMessage request from %s: %#v", from, r)
		return true, nil
	}

	b.Log().Info("received unknown request: %#v", request)
	return false, nil
}

```

go.mod
```
module demo

go 1.21.6

require (
	ergo.services/ergo v1.999.225-0.20240303205107-3b21499342ce
	ergo.services/logger v0.0.0-20240221211214-98de4c9ff50e
)

require (
	github.com/fatih/color v1.16.0 // indirect
	github.com/mattn/go-colorable v0.1.13 // indirect
	github.com/mattn/go-isatty v0.0.20 // indirect
	golang.org/x/sys v0.14.0 // indirect
)

```
result:
![image](https://github.com/ergo-services/ergo/assets/118860/7abc515f-c4d8-4d4f-957d-ba947e6ce983)

**Reply by Northward712** [reply]
Thank you very much for this demo, and I tried running it. It works perfectly fine. However, in my testing environment, I didn’t start two nodes in the same process. Instead, I started one Node in each of two different processes to make an inter-process call. In this setup, actor.Send works without any issues, but actor.Call encounters problems. The names of the two nodes are 10001@127.0.0.1 and 10003@127.0.0.1. I’m not sure if the issue arises from how I started the two node processes or something else. Thank you for your response.

**Reply by halturin** [reply]
it is not a problem to split it out into two different modules, but make sure to register the message type (`edf.RegisterTypeOf(...)`) of your message on both sides (if this message is declared in the module of node1, you should import it and register exactly this type).

**Reply by Northward712** [reply]
I simply used the string type.

**Comment by halturin** [comment]
also added this example here https://github.com/ergo-services/examples/tree/v300

**Comment by Northward712** [comment]
![截屏2024-03-05 22 52 11](https://github.com/ergo-services/ergo/assets/37608723/b407e5f5-6386-4626-a8ab-2b46ce5c5068)
![截屏2024-03-05 22 53 32](https://github.com/ergo-services/ergo/assets/37608723/fe632453-d47c-41cb-98ae-03aca39d5330)
![截屏2024-03-05 22 53 48](https://github.com/ergo-services/ergo/assets/37608723/bbf9a12d-b2d0-47ba-8766-a2ca7bd328a0)

This is my calling process. As you can see, when HandleCall actively throws an error, it will be returned to the caller. However, at this point, the caller’s Call is still blocked until it times out. In other words, there is no issue with network connectivity or the logic of the Call; it’s just unclear why the correct response is not being received.

**Comment by Northward712** [comment]
![截屏2024-03-05 22 56 19](https://github.com/ergo-services/ergo/assets/37608723/d1123da8-d0b2-42d4-b154-f18b3f6eb747)
 
![截屏2024-03-05 22 58 14](https://github.com/ergo-services/ergo/assets/37608723/20cf0fbc-11c5-4d2e-9aff-a9f35847ac88)

This is the code that does not actively throw an error. As you can see in the console, it did not return “try reply remote call” as expected, but instead threw a timeout.

**Comment by halturin** [comment]
Just fixed an issue that might be related to your problem. Pull the latest changes in v300.

**Comment by davidroman0O** [comment]
I've been reading and playing with the v3 examples, those meta processes injected on the endpoint to then use a pool (inited on the supervisor) then a web worker that handle the methods... it's so clever! I love it!

<img width="734" alt="Screenshot 2024-07-06 at 10 26 35 AM" src="https://github.com/ergo-services/ergo/assets/2845356/63708529-b21c-4e4d-b118-0cd2131501de">

**Reply by halturin** [reply]
thanks. 
PS just noticed some misprints in the comments :). will fix them before the release
